vault/expiration.go

```diff
@@ -422,6 +480,88 @@ func (m *ExpirationManager) Restore(errorFunc func()) (retErr error) {
 	m.restoreModeLock.Unlock()
 
 	m.logger.Info("lease restore complete")
+
+	if atomic.LoadUint32(&m.SHA1HashedLeasesCleared) == 0 {
+		expMetadata := &ExpirationMetadata{}
+		if atomic.LoadUint32(&m.SHA1HashedUnleasedTokensUpgraded) == 0 {
+			// The obfuscated tokens remaining in the map after lease restoration
+			// would only be those of root tokens without leases and of tokens that
+			// have lost leases. Revoke all the tokens that have lost the leases
+			// and upgrade the root tokens to use SHA2-256 HMAC instead of SHA1
+			// hash.
+			for obfuscatedToken, _ := range m.obfuscatedTokens {
+				// lookupObfuscatedToken will take care of revoking tokens that
+				// do not have an associated lease. The token entries that get
+				// returned should only be of root tokens.
+				te, err := m.tokenStore.lookupObfuscatedToken(m.quitContext, obfuscatedToken, false)
+				if err != nil {
+					return err
+				}
+
+				// For sanity
+				if te == nil {
+					continue
+				}
+
+				// If the token entry is of an older version, perform the upgrade
+				if te.Version < 2 {
+					originalNumUses := te.NumUses
+					// Calling token store's revoke function instead of removing
+					// the storage entries manually has the advantage of not
+					// missing out on removal of related artifacts such as the
+					// accessor, cubbyhole and parent index. The disadvantage
+					// however is that the lookup is performed again. Since this
+					// upgrade happens only once, it should be okay.
+					err = m.tokenStore.revokeObfuscatedToken(m.quitContext, obfuscatedToken, false, te)
```

Comment:
I wonder if instead of revoking the token if we should instead just be updating the storage key names. Revoking could have unintended consequences of revoking a lot of leases and child tokens. Though i wonder if we'd need to update the storage entries for all the children tokens and leases. Which is also not ideal...

Reply:
Yeah, I don't think revoking here is the right thing to do. Instead we should probably just go through and update all relevant entries (and only remove the old ones once the new ones are persisted successfully).

Reply:
Okay. This would mean that when a upgrade candidate is encountered, we should upgrade the storage entries for cubbyhole, accessor, parent index and the token itself. When all this is successful, we can delete the old entries. Considering that the format of the new cubbyhole entries are different, I wonder if upgrading all the cubbyhole items can be tricky.

I am not sure if we want to upgrade the storage entries for the children though. They fall to the category where `SHA1HashedLeasesCleared` tracks the presence of such leases.

Reply:
I wonder if we could upgrade only the token value itself and leave the rest intact. Something like change the actual lookup value to be hmac'd but keep the version at 0 (or maybe even: version 1). The accessor is sensitive but not secret, the cubbyhole ID is double-hashed anyways.